data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java

```diff
@@ -1008,7 +996,31 @@ public void bindOne(@NonNull QueryParameterBinding binding, Object value) {
 
                 @Override
                 public void bindMany(@NonNull QueryParameterBinding binding, @NonNull Collection<Object> values) {
-                    bindOne(binding, values);
+                    // For IN/NOT IN criteria Cosmos won't return results for IN (@p1) because it expects IN (@p1, @p2,...)
+                    // so we must expand parameters and update query with newly created parameters
+                    String parameterName = getParameterName(binding, isRawQuery);
+                    // This should find if there is IN (@p1) or NOT IN (@p1) which may or may not have spaces in between
+                    Pattern inPattern = Pattern.compile("IN\\s*\\(\\s*" + ("@" + parameterName) + "\\s*\\)", Pattern.CASE_INSENSITIVE);
```

Comment:
There is a lot of logic to prevent runtime query evaluation. Fo the SQL the binding can be expandable, which means we will store the query as parts, and at the runtime, we can create a query without analyzing which part should be split.

Reply:
Ok, I wasn't aware this implementation was missing to call prepareQuery that does it. I have updated it here https://github.com/micronaut-projects/micronaut-data/pull/1813/commits/8752d547fd602fe9e892e7624c9f624fff73abe0
Had to override prepareQuery because of positional parameters for IN/NOT IN and smaller change in bind logic but this looks way better than prev version. Thanks